Fix `with_spectral_unit`

rosteen opened the pull request:
This has been needed for a long time - intended to fix #995 , #889 , #676. My current changes fix the behavior for a `Spectrum1D` initialized with a `spectral_axis`, but I think it needs a little more work to also fix it for one initialized with a `wcs`. Note that this also moves the `SpectralGWCS` out of `gwcs_from_array` so that it's actually accessible to check instances, since there are times we may want to do something different if that's the WCS type.

rosteen commented:
Ok, I realized the existing WCS-initialized spectrum case was just acting on the `spectral_axis` as well, so I didn't need the a special case for the `spectral_axis`-initialized case and handle both the same way. It would probably be more ideal to preserve as much of the original WCS as possible while changing the unit, but that can be work for another day. This at least restores the previous (intended) functionality.

rosteen commented on specutils/utils/wcs_utils.py:
```diff
@@ -8,6 +8,18 @@
+        return super().pixel_to_world(*args, **kwargs).to(
+            self.original_unit, equivalencies=u.spectral())
+
+
```
Weird, I thought I ran the tests locally after adding that, guess not. Thanks for the link.

rosteen replied:
Actually the tests do pass locally, I must have the IERS files cached.

rosteen commented:
Looks like the remaining failures are a (hopefully transient) download failure:

```
astropy.utils.iers.iers.IERSWarning: failed to download https://datacenter.iers.org/data/9/finals2000A.all: <urlopen error [Errno -3] Temporary failure in name resolution>
```

dhomeier commented:
Yes, seen those frequently, although they did not show up in the previous run; unlikely this was introduced by adding the copy methods. Can maybe check if astropy has done something to avoid these connection failures.

dhomeier commented:
Seems the last solution was just to wait till DNS is working again... astropy/astropy#15456

dhomeier commented on specutils/spectra/spectrum_mixin.py:
```diff
@@ -208,14 +215,27 @@ def with_spectral_unit(self, unit, velocity_convention=None,
+            meta['original_unit'] = orig_unit
+
+        if 'original_wcs' not in self.meta:
+            meta['original_wcs'] = self.wcs.copy()
```
```suggestion
            meta['original_wcs'] = self.wcs.deepcopy()
```
Since this brought up the issue of the convenience copy method, you might actually want this, if you really treat the  created spectrum as a new object; otherwise any changes to `new_spec.wcs` will modify the original as well.
Should perhaps be added to the test below for an instance with wcs.

rosteen replied:
Good call, I can never keep straight when copy vs deepcopy is the right answer.

rosteen commented:
@dhomeier I switched to `deepcopy` and added a test to check that the stored copy of the original WCS has the original unit. If this looks good to you now would you approve so I can merge? Thanks!

dhomeier commented:
The test I did locally to verify it really needs `deepcopy` was (kinda brute force) setting `new_spec.wcs.cdelt` (with an Astropy WCS) to different values and check if the original was changed as well. But `SpectralGWCS` might not be modified that easily, so maybe that's overkill.

rosteen commented:
> check if the original `spec.wcs.cdelt` was changed by this

Doesn't checking that the `world_axis_units` is different between the original WCS and new WCS confirm that they are different objects? I think this test is accomplishing the same thing as yours.

> Just noted that the entire `meta` is copied as well; being a dictionary, this would probably need a `deepcopy` as well to ensure you have separate objects.

The meta is an `OrderedDict`, which doesn't seem to have `deepcopy`. I just confirmed that `meta is self._meta` returns False after the `copy`, so I think this is ok.

dhomeier commented on specutils/spectra/spectrum_mixin.py:
```diff
@@ -208,14 +215,27 @@ def with_spectral_unit(self, unit, velocity_convention=None,
+                                                 rest_value)
+
+        # Store the original unit information and WCS for posterity
+        meta = self._meta.copy()
```
```suggestion
        meta = self._meta.deepcopy()
```
This would also be needed if `meta` itself includes nested objects (see test example)

dhomeier commented on specutils/tests/test_spectrum1d.py:
```diff
@@ -150,10 +150,17 @@ def test_spectral_axis_conversions():
+    assert new_spec.wcs.world_axis_units[0] == "km.s**-1"
+    # Make sure meta stored the old WCS correctly
+    assert new_spec.meta["original_wcs"].world_axis_units[0] == "nm"
+    assert new_spec.meta["original_spectral_axis_unit"] == "nm"
```
```suggestion
    assert new_spec.meta["original_spectral_axis_unit"] == "nm"

    wcs_dict = {"CTYPE1": "WAVE", "CRVAL1": 3.622e3, "CDELT1, "CDELT3": 8e-2, "CRPIX1": 0, "CUNIT1": "Angstrom"}
    wcs_spec = Spectrum1D(flux=np.random.randn(49) * u.Jy, wcs=WCS(wcs_dict),
                          meta={'header': wcs_dict.copy()})
    new_spec = wcs_spec.with_spectral_axis_unit(u.km/u.s, rest_value=125*u.um,
                                       velocity_convention="relativistic")
    new_spec.wcs.wcs.crval = [3777]
    new_spec.meta['header']['CRVAL1'] = 3777.0

    assert wcs_spec.wcs.wcs.crval[0] == 3.622e-7
    assert wcs_spec.meta['header']['CRVAL1'] == 3622.
```
This should provide a basic test that the `new_spec` components are independent and not propagating changes back to `wcs_spec` (the second one should require the `self._meta.deepcopy()` change above).

dhomeier reviewed (commented):
Tried to set up an example of the deepcopy tests I commented on

rosteen commented:
@dhomeier I did update the docstring already to explain what happens to the original WCS, is that enough to address the concern you brought up about documentation earlier?

dhomeier commented on specutils/spectra/spectrum_mixin.py:
```diff
@@ -184,10 +182,19 @@ def velocity(self):
         """
-        Returns a new spectrum with a different spectral axis unit.
+        Returns a new spectrum with a different spectral axis unit. Note that this creates a new
+        object using the converted spectral axis and thus drops the original WCS, if it existed.
+        The original WCS will be stored in the ``original_wcs`` entry of the new object's ``meta``
+        dictionary.
```
```suggestion
        object using the converted spectral axis and thus drops the original WCS, if it existed,
        replacing it with a :class:`~specutils.utils.wcs_utils.SpectralGWCS`. The original WCS
        will be stored in the ``original_wcs`` entry of the new object's ``meta`` dictionary.
```
That's about my only point left about the documentation, to make it clear that `new_spec.wcs` will be a different class of object. Maybe `SpectralGWCS` could also use a basic docstring, if only to link to `~gwcs.WCS`...

dhomeier reviewed (approved):
Thanks; leaving some final remarks on the docs, but this looks good now!

dhomeier commented:
Not sure – do we get
> WARNING: py:class reference target not found: specutils.utils.wcs_utils.SpectralGWCS

because that version is not in the online docs yet, or actually because `SpectralGWCS` does not have a docstring yet?

But I am seeing other [broken links](https://specutils.readthedocs.io/en/latest/api/specutils.analysis.moment.html#specutils.analysis.moment) into `utils` that apparently have not broken the build.

dhomeier commented on specutils/analysis/moment.py:
```diff
@@ -21,7 +21,7 @@ def moment(spectrum, regions=None, order=0, axis=-1):
         The spectrum object over which the width will be calculated.
 
-    regions: `~specutils.utils.SpectralRegion` or list of `~specutils.utils.SpectralRegion`
+    regions: `~specutils.SpectralRegion` or list of `~specutils.SpectralRegion`
```
```suggestion
    regions: `~specutils.spectra.spectral_region.SpectralRegion` or list of `~specutils.spectra.spectral_region.SpectralRegion`
```
looks like these need the full path, even if they are exposed at the top level (cf. Spectrum1D reference above).

dhomeier commented on specutils/analysis/uncertainty.py:
```diff
@@ -22,7 +22,7 @@ def snr(spectrum, region=None):
         The spectrum object overwhich the equivalent width will be calculated.
 
-    region: `~specutils.utils.SpectralRegion` or list of `~specutils.utils.SpectralRegion`
+    region: `~specutils.SpectralRegion` or list of `~specutils.SpectralRegion`
```
```suggestion
    region: `~specutils.spectra.spectral_region.SpectralRegion` or list of `~specutils.spectra.spectral_region.SpectralRegion`
```

dhomeier commented on specutils/analysis/uncertainty.py:
```diff
@@ -67,7 +67,7 @@ def _snr_single_region(spectrum, region=None):
         The spectrum object overwhich the equivalent width will be calculated.
 
-    region: `~specutils.utils.SpectralRegion`
+    region: `~specutils.SpectralRegion`
```
```suggestion
    region: `~specutils.spectra.spectral_region.SpectralRegion`
```

dhomeier commented on specutils/analysis/uncertainty.py:
```diff
@@ -109,7 +109,7 @@ def snr_derived(spectrum, region=None):
         The spectrum object overwhich the equivalent width will be calculated.
 
-    region: `~specutils.utils.SpectralRegion`
+    region: `~specutils.SpectralRegion`
```
```suggestion
    region: `~specutils.spectra.spectral_region.SpectralRegion`
```

dhomeier commented on specutils/analysis/uncertainty.py:
```diff
@@ -155,7 +155,7 @@ def _snr_derived(spectrum, region=None):
         The spectrum object overwhich the equivalent width will be calculated.
 
-    region: `~specutils.utils.SpectralRegion`
+    region: `~specutils.SpectralRegion`
```
```suggestion
    region: `~specutils.spectra.spectral_region.SpectralRegion`
```

dhomeier reviewed (commented):
Original problem solved, new barrel opened!
At least the docs are getting a good overhaul now. :D

rosteen commented:
@dhomeier After some tinkering I couldn't get that doc link to work (I was breaking more things...) so I decided to just make the docstring a little more descriptive here for now as a compromise. I pointed those other `utils` doc links to the proper place, but they still seem to be broken rather than linking 🤔. I'll come back to that at some point but I have more pressing concerns at the moment.

rosteen commented:
@dhomeier As I commented above on one of your suggestions, you can see [here](https://specutils--1119.org.readthedocs.build/en/1119/api/specutils.analysis.moment.html#specutils.analysis.moment) that using the full path to `SpectralRegion` doesn't fix the broken link. If you want to debug that and open a follow-up PR that would be awesome, but I don't think it should hold up this one.

dhomeier commented:
Sorry, was a bit too fast in committing then – maybe you should force-push again to revert.
Sphinx keeps puzzling me; I just noted that in the docs themselves the recommended path is
```python
from specutils.spectra import SpectralRegion
```
but this can wait for another PR as well.
